Libraries/LibJS/Runtime/DatePrototype.cpp

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2020, Linus Groh <mail@linusgroh.de>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/Function.h>
+#include <AK/String.h>
+#include <LibCore/DateTime.h>
+#include <LibJS/Interpreter.h>
+#include <LibJS/Runtime/Date.h>
+#include <LibJS/Runtime/DatePrototype.h>
+#include <LibJS/Runtime/Value.h>
+
+namespace JS {
+
+DatePrototype::DatePrototype()
+{
+    put_native_function("getDate", get_date);
+    put_native_function("getDay", get_day);
+    put_native_function("getFullYear", get_full_year);
+    put_native_function("getHours", get_hours);
+    put_native_function("getMilliseconds", get_milliseconds);
+    put_native_function("getMinutes", get_minutes);
+    put_native_function("getMonth", get_month);
+    put_native_function("getSeconds", get_seconds);
+    put_native_function("getTime", get_time);
+}
+
+DatePrototype::~DatePrototype()
+{
+}
+
+Value DatePrototype::get_date(Interpreter& interpreter)
+{
+    auto* this_object = interpreter.this_value().to_object(interpreter.heap());
+    if (!this_object)
+        return {};
+    ASSERT(this_object->is_date());
```

Comment:
Let's make a helper function for turning `this` into a `DateObject`, something like this:

```cpp
DateObject* this_date_from_interpreter(Interpreter& interpreter)
{
    auto* this_object = interpreter.this_value().to_object(interpreter.heap());
    if (!this_object)
        return nullptr;
    if (!this_object->is_date()) {
        interpreter.throw_exception<Error>("TypeError", "this is not a DateObject");
        return nullptr;
    }
    return static_cast<DateObject*>(this_object);
}
```

WIth that, all the functions can call it up front:

```cpp
auto* this_object = this_date_from_interpreter(interpreter);
if (!this_object)
    return {};
...
```